### PR TITLE
Series datetime is_year_end and is_year_start

### LINKED ISSUE #8680: [FEA] Series datetime is_year_end and is_year_start
For feature engineering tasks, I'd like to be able to determine whether a datetime is the beginning or end of a year, like I can in pandas.

```python
import pandas as pd
​
s = pd.Series(["2021-02-27", "2020-03-31"], dtype="datetime64[ms]")
s.dt.is_year_end
0    False
1    False
dtype: bool
```

```python
import pandas as pd
​
s = pd.Series(["2021-01-01", "2020-04-01"], dtype="datetime64[ms]")
s.dt.is_year_start
0     True
1    False
dtype: bool
```

### PR DESCRIPTION
**Pull request by marlenezw**
This PR aims to allow users to be able to determine whether a datetime is the beginning or end of a year. This PR closes #8680

### CODE REVIEW

File: python/cudf/cudf/core/series.py [hunk 1]
```diff
@@ -6448,6 +6448,30 @@ def is_month_start(self):
         """
         return (self.day == 1).fillna(False)
 
+    @property
+    def is_year_start(self):
+        """
+        Boolean indicator if the date is the first day of the month.
```
Comment (skirui-source) [inline python/cudf/cudf/core/series.py:6454]: typo fix: i think you meant if the date is the first day of year

File: python/cudf/cudf/core/series.py [hunk 2]
```diff
@@ -6448,6 +6448,30 @@ def is_month_start(self):
...
+        -------
+        Series
+        Booleans indicating if dates are the first day of the month.
+        """
+        return (self.day_of_year == 1).fillna(False)
+
+    @property
+    def is_year_end(self):
+        """
+        Boolean indicator if the date is the first day of the month.
```
Comment (skirui-source) [inline python/cudf/cudf/core/series.py:6466]: similar typo-fix:  indicates if date is the end of year

File: python/cudf/cudf/core/series.py [hunk 3]
```diff
@@ -6448,6 +6448,30 @@ def is_month_start(self):
...
+
+        Returns
+        -------
+        Series
+        Booleans indicating if dates are the first day of the month.
+        """
+        return (self.day_of_year == 365).fillna(False)
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/series.py:6473]: does this work for leap years?

File: python/cudf/cudf/core/series.py (hunk 1 above)
Comment (marlenezw) [reply python/cudf/cudf/core/series.py:6454]: Thanks for this catch!

File: python/cudf/cudf/core/series.py (hunk 3 above)
Comment (marlenezw) [reply python/cudf/cudf/core/series.py:6473]: You're right, this does not work for leap years. Adding to the code to factor that in.

**Comment by marlenezw** [comment]
rerun tests

File: python/cudf/cudf/core/series.py [hunk 4]
```diff
@@ -6006,6 +6006,32 @@ def days_in_month(self):
...
+        -------
+        Series
+        Booleans indicating if dates are the last day of the year.
+        """
+        leap_end_days = (self.day_of_year == 366).fillna(False)
+        if self.is_leap_year.any() and leap_end_days.any():
+            year_end_days = (self.day_of_year == 365).fillna(False)
+            return leap_end_days + year_end_days
+        return (self.day_of_year == 365).fillna(False)
+
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/series.py:6034]: s is:
```python
In [5]: s = cudf.Series(
   ...: [
   ...:             "2020-12-31 08:00:00",
   ...:             None,
   ...:             "1999-12-31 18:40:00",
   ...:             "2000-12-31 04:00:00",
   ...:             None,
   ...:             "1800-12-14 07:30:00",
   ...:             "2100-12-14 07:30:00",
   ...:             "1970-01-01 00:00:00",
   ...:             "1969-12-31 12:59:00",
   ...:             "2020-05-31",
   ...:             None,
   ...:             "1999-12-01",
   ...:             "2000-12-21",
   ...:             None,
   ...:             "1900-12-31",
   ...:             "2017-12-30",
   ...:             "2017-12-31",
   ...:             "2020-05-31",
   ...:             None,
   ...:             "1999-12-01",
   ...:             "2000-12-21",
   ...:             None,
   ...:             "1900-12-31",
   ...:             "1800-03-14",
   ...:             "2100-03-10",
   ...:             "1970-01-01",
   ...:             "1969-12-11",
   ...:             "2018-01-01",
   ...:             "2017-12-30",
   ...:             "2017-12-31",
   ...:         ]*100_00_0, dtype='datetime64[ms]')

In [6]: s
Out[6]: 
0          2020-12-31 08:00:00.000
1                             <NA>
2          1999-12-31 18:40:00.000
3          2000-12-31 04:00:00.000
4                             <NA>
                    ...           
2999995    1970-01-01 00:00:00.000
2999996    1969-12-11 00:00:00.000
2999997    2018-01-01 00:00:00.000
2999998    2017-12-30 00:00:00.000
2999999    2017-12-31 00:00:00.000
Length: 3000000, dtype: datetime64[ms]
```

[the remark this answers is not in the document]
Comment (marlenezw) [reply python/cudf/cudf/core/series.py:6040]: This is super helpful Prem! Thanks for the feedback!

File: python/cudf/cudf/core/series.py [hunk 5]
```diff
@@ -6006,6 +6006,32 @@ def days_in_month(self):
...
 
+    @property
+    def is_year_start(self):
+        """
+        Boolean indicator if the date is the first day of the year.
+        Returns
+        -------
+        Series
+        Booleans indicating if dates are the first day of the year.
+        """
```
[the remark this answers is not in the document]
Comment (marlenezw) [reply python/cudf/cudf/core/series.py:6017]: Sounds good 😄

File: python/cudf/cudf/tests/test_datetime.py [hunk 6]
```diff
@@ -1346,3 +1346,100 @@ def test_is_month_start(data, dtype):
...
+            "2000-12-21",
+            None,
+            "1900-12-31",
+            "2017-12-30",
+            "2017-12-31",
+        ]
+    ],
+)
+@pytest.mark.parametrize("dtype", ["datetime64[ns]"])
+def test_is_year_end_leap_year(data, dtype):
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_datetime.py:1438]: Can we just remove this pytest and have the data added into `test_is_year_end`

Comment (marlenezw) [reply python/cudf/cudf/tests/test_datetime.py:1438]: Sounds good too!

File: python/cudf/cudf/core/series.py [hunk 7]
```diff
@@ -6033,6 +6033,76 @@ def days_in_month(self):
...
+        0    False
+        1    False
+        2    True
+        dtype: bool
+        """
+        outcol = self.series._column.get_dt_field(
+            "day_of_year"
+        ) == cudf.Scalar(1)
+        return Series._from_data(
+            ColumnAccessor({None: outcol.fillna(False)}),
```
Comment (galipremsagar) [inline python/cudf/cudf/core/series.py:6064]:
```suggestion
            {None: outcol.fillna(False)},
```

File: python/cudf/cudf/core/series.py [hunk 8]
```diff
@@ -6033,6 +6033,76 @@ def days_in_month(self):
...
+        """
+        day_of_year = self.series._column.get_dt_field("day_of_year")
+        leap_dates = libcudf.datetime.is_leap_year(self.series._column)
+
+        leap = day_of_year == cudf.Scalar(366)
+        non_leap = day_of_year == cudf.Scalar(365)
+        result = cudf._lib.copying.copy_if_else(leap, non_leap, leap_dates)
+        result = result.fillna(False)
+        return Series._from_data(
+            ColumnAccessor({None: result}),
```
Comment (galipremsagar) [inline python/cudf/cudf/core/series.py:6101]:
```suggestion
           {None: result},
```

**Review by galipremsagar — CHANGES_REQUESTED** [review]

**Review by galipremsagar — APPROVED** [review]

**Review by brandon-b-miller — APPROVED** [review]

**Comment by marlenezw** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**